### ISSUE TITLE
Error in tune mode

### ISSUE THREAD

**Issue by nikopartanen** [pyannote/pyannote-audio#39]
I get following error messages when trying to run the tune command on my adapted template. With training everything went as expected, I think. This creates `tune.png` and `tune.yml` files, but I assume this doesn't look as it should:

![Plot](http://i.imgur.com/c1OuhTO.png)

This is the error message:

    Iteration No: 1 ended. Evaluation done at provided point.
    Time taken: 3.3038
    Function value obtained: 0.0000
    Current minimum: 0.0000
    Iteration No: 2 started. Evaluating function at random point.
    /Users/niko/anaconda3/envs/py35-pyannote-audio/lib/python3.5/site-packages/keras/models.py:245: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.
      warnings.warn('No training configuration found in save file: '
    /Users/niko/anaconda3/envs/py35-pyannote-audio/lib/python3.5/site-packages/skopt/optimizer/optimizer.py:195: UserWarning: The objective has been evaluated at this point before.
      warnings.warn("The objective has been evaluated "
      ... (the same message repeated)
      Iteration No: 20 ended. Search finished for the next optimal point.
    Time taken: 8.3426
    Function value obtained: 0.0000
    Current minimum: 0.0000
    Exception ignored in: <bound method BaseSession.__del__ of <tensorflow.python.client.session.Session object at 0x12d522a58>>
    Traceback (most recent call last):
      File "/Users/niko/anaconda3/envs/py35-pyannote-audio/lib/python3.5/site-packages/tensorflow/python/client/session.py", line 582, in __del__
    AttributeError: 'NoneType' object has no attribute 'TF_DeleteStatus'

**Comment by hbredin** [comment]
As far as the convergence plot is concerned, this is indeed not good.
By default, "tune" mode relies on the "development" subset of the provided protocol.
I suspect your "development" set is empty. Does your protocol contain a "development" set?

As far as the "tensorflow" error is concerned, I have to look into it. 
I've not been using "tensorflow" myself, but used "theano" backend up to now...

**Comment by nikopartanen** [comment]
I noticed now that I hadn't specified those sets! I will try to fix that now first.

**Comment by hbredin** [comment]
To use "theano" backend, you need to set `KERAS_BACKEND` environment variable to `"theano"`.

Though it is not recommended, you can tune your model on the training set itself, by using the option `--subset=train`.

**Comment by nikopartanen** [comment]
Thanks, setting `KERAS_BACKEND` seems to work. Is there somewhere a template where `train`, `test` and `dev` subsets are all filled? So there should be three different `.mdtm` files? In `__init__.py` there is now:

    def dev_iter(self):
        # here, you should do the same as above, but for the development set
        for _ in []:
            yield

But I'm not sure how these should be defined.

**Comment by hbredin** [comment]
Yes. Three different files (one for train, one for dev, one for test).
`trn_iter` should iterate on `train` files, `dev_iter` on `development` files, and `tst_iter` on `test`files.

This dataset split is kind of standard: `train` is used for training your model, `development` is used to tune hyper parameters (e.g. thresholds) and `test` is used to evaluate the whole approach on unseen data.
 
I am not sure this answers your question, though.

**Comment by nikopartanen** [comment]
This helped quite a bit! Or at least the problem changed. [Here](https://github.com/langdoc/pyannote-db-ikdp/blob/master/ikdp/__init__.py) is how I have defined the development and test sets now, it may be the problem is still there in way how it is set up.

Now I get in tuning part a different error:

    Using Theano backend.
    Iteration No: 1 started. Evaluating function at provided point.
    /Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/keras/models.py:245: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.
      warnings.warn('No training configuration found in save file: '
    /Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/pyannote/database/util.py:193: UserWarning: "annotated" was approximated by "wav" duration.
      warnings.warn('"annotated" was approximated by "wav" duration.')
    /Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/pyannote/database/util.py:193: UserWarning: "annotated" was approximated by "wav" duration.
      warnings.warn('"annotated" was approximated by "wav" duration.')
    /Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/pyannote/database/util.py:193: UserWarning: "annotated" was approximated by "wav" duration.
      warnings.warn('"annotated" was approximated by "wav" duration.')
    /Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/pyannote/database/util.py:193: UserWarning: "annotated" was approximated by "wav" duration.
      warnings.warn('"annotated" was approximated by "wav" duration.')
    multiprocessing.pool.RemoteTraceback: 
    """
    Traceback (most recent call last):
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/multiprocessing/pool.py", line 119, in worker
        result = (True, func(*args, **kwds))
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/multiprocessing/pool.py", line 44, in mapstar
        return list(map(*args))
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/pyannote/audio/signal.py", line 132, in helper_binarize_tune
        hypothesis_timeline = binarizer.apply(prediction, **kwargs)
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/pyannote/audio/signal.py", line 313, in apply
        active = active.coverage()
    AttributeError: 'Timeline' object has no attribute 'coverage'
    """
    
    The above exception was the direct cause of the following exception:
    
    Traceback (most recent call last):
      File "/Users/niko/anaconda3/envs/pyannote/bin/pyannote-speech-detection", line 11, in <module>
        sys.exit(main())
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/pyannote/audio/applications/speech_detection.py", line 607, in main
        application.tune(protocol_name, subset=subset)
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/pyannote/audio/applications/speech_detection.py", line 499, in tune
        verbose=True, callback=callback)
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/skopt/optimizer/gp.py", line 247, in gp_minimize
        callback=callback, n_jobs=n_jobs)
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/skopt/optimizer/base.py", line 216, in base_minimize
        y0 = list(y0)
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/pyannote/audio/applications/speech_detection.py", line 488, in objective_function
        self, epoch, protocol_name, subset=subset)
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/pyannote/audio/applications/speech_detection.py", line 221, in tune_binarizer
        dimension=1)
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/pyannote/audio/signal.py", line 252, in tune
        random_state=1337, verbose=False)
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/skopt/optimizer/gp.py", line 247, in gp_minimize
        callback=callback, n_jobs=n_jobs)
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/skopt/optimizer/base.py", line 237, in base_minimize
        next_y = func(next_x)
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/site-packages/pyannote/audio/signal.py", line 239, in objective_function
        zip(items, predictions)))
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/multiprocessing/pool.py", line 260, in map
        return self._map_async(func, iterable, mapstar, chunksize).get()
      File "/Users/niko/anaconda3/envs/pyannote/lib/python3.5/multiprocessing/pool.py", line 608, in get
        raise self._value
    AttributeError: 'Timeline' object has no attribute 'coverage'

This happens also when I use `--subset=train`. It doesn't create anything into `tune/ikdp.SpeakerDiarization.MyFirstProtocol.development` folder.

**Comment by hbredin** [comment]
This bug has been fixed in `develop` branch here: 
https://github.com/pyannote/pyannote-audio/commit/779eb8dc54f56c26faa211d694f2a720ce15931b

I haven't done a proper new release yet, though. 
In the meantime, I suggest you use the `develop` branch:

```
$ git clone https://github.com/pyannote/pyannote-audio.git
$ cd pyannote-audio
$ git checkout develop
$ pip install -e .
```

**Comment by nikopartanen** [comment]
Now I was able to go through [LSTM-based speech activity detection tutorial](https://github.com/pyannote/pyannote-audio/tree/master/tutorials/speech-activity-detection) without problems in tuning or applying. I just finished testing those, and will now evaluate the results a bit and repeat it with a larger dataset and longer training, now I stopped it after 25 epochs. Thanks a lot for help!

My tune.png plot is still very flat, but at least isn't in zero as earlier.

![](http://i.imgur.com/AryGugw.png)
